test/unit/nested.cpp

```diff
@@ -0,0 +1,31 @@
+#include "gtest/gtest.h"
+
+#include "RAJA/pattern/nested.hpp"
+
+#include <iostream>
+
+TEST(Nested, Basic)
+{
+  using namespace RAJA::nested;
+  using Index_type = RAJA::Index_type;
+  using pol = Policy<For<1, RAJA::seq_exec>, For<0, RAJA::seq_exec>>;
+  RAJA::nested::forall(pol{},
+                       RAJA::util::make_tuple(RAJA::RangeSegment(0, 5),
+                                              RAJA::RangeSegment(0, 5)),
+                       [=](Index_type i, Index_type j) {
+                         std::cout << i << " " << j << std::endl;
+                       });
+}
+
+TEST(Nested, collapse)
+{
+  using namespace RAJA::nested;
+  using Index_type = RAJA::Index_type;
+  using pol = Policy<Collapse<RAJA::seq_exec, CFor<1>, CFor<0>>>;
```

Comment:
As someone who honestly struggled with forallN a lot last summer, I already see substantial improvement. The biggest one for me is writing things out the way you want them to happen. With forallN tiling and permutation didn't always make the most sense. You had to imagine the structure by applying them in your head first.